Draft dockerization

yakky opened the pull request:
Dockerizzazione iniziale del progetto.

Attualmente assume un ambiente di sviluppo per alcuni elementi (dominio, p.es) che andranno poi cambiati prima della finalizzazione

La struttura attuale risente pesantemente dell'attuale layout e architettura del deployment monolitico su virtual machine (sopratutto la divisione web/api e fra i diversi task celery e relativa concurrency), probabilmente può essere opportuno lavorare in 2 fasi:

1) replicare in container l'architettura attuale con miglioramenti
2) fare una vera ottimizzazione

La fase 2 probabilmete ha più senso dopo l'upgrade del progetto (#251 ) per verificare i cambiamenti nelle versioni moderne di RTD

Problemi da risolvere:

- [x] la ricerca non ritorna risultati #273 
- [x] fallisce la build local #274 
- [x] gestione dei secret #276 
- [x] inizializzazione ES #277
- [x] Completare il [Setup dell'ambiente di sviluppo locale](https://github.com/italia/docs.italia.it/wiki/[DRAFT]-Setup-dell'ambiente-di-sviluppo-locale)

yakky committed:
Dockerization

yakky committed:
Dockerization

francescozaia commented on .env:
```diff
@@ -0,0 +1,45 @@
+RTD_USER=docs
+DEBUG=1
+DOCS_BASE=/home/documents
+ACCOUNT_HTTP_PROTO=http
```
A naso `SLUMBER_PASSWORD`, `GITHUB_API_SECRET` ed altre potenziali come `EMAIL_HOST_PASSWORD` o `SECRET_KEY` non dovrebbero stare qui, o sbaglio?

yakky replied:
si, hai perfettamente ragione, errore da niubbo terribile
ho fatto un force per cancellarne le tracce

yakky replied:
rimosso tutto

francescozaia commented on docker/nginx/snippets/docs-redirects.conf:
```diff
@@ -0,0 +1,4 @@
+# Document "Piano Triennale" - move permanently version "stabile" to version "2017-2019"
+# This is to prevent broken links pointing to version "stabile"
+# The version "stabile" needs to be deactivated
+rewrite ^/italia/piano-triennale-ict/pianotriennale-ict-doc/it/stabile/(.*)$ /italia/piano-triennale-ict/pianotriennale-ict-doc/it/2017-2019/$1 permanent;
```
Questo codice è nuovo o è già presente altrove al momento?

yakky replied:
questa è la configurazione nginx attualmente in produzione

francescozaia commented on readthedocs/docsitalia/settings/docker.py:
```diff
@@ -0,0 +1,259 @@
+        # pylint: disable=unused-wildcard-import
+        from .local_settings import *  # noqa
+    except ImportError:
+        pass
```
Nota di carattere generale: commentiamo molto di più i blocchi di variabili, in modo che sia chiaro a chi ci dovrà lavorare in futuro gli scopi di ciascuna di esse.

yakky replied:
approvo
quel file di settings va completamente riscritto e documentato

francescozaia commented on readthedocs/templates/doc_builder/conf.py.tmpl:
```diff
@@ -38,7 +38,7 @@ else:
 # Support markdown
 def setup(app):
-    app.add_source_parser('.md', CommonMarkParser)
+#     app.add_source_parser('.md', CommonMarkParser)
```
Probabilmente non è pertinente, ma giusto per essere allineati: vorremmo mantenere il supporto a Markdown per le documentazioni.

yakky replied:
Giusto, è un quickfix che serve quando la build dei documenti avviene fuori dal container docker di build

Quel codice sarà comunque da rimuovere durante l'aggiornamento perché le nuove versioni di rtd supportando markdown nativamente

yakky replied:
rimosso

francescozaia commented on readthedocs/docsitalia/settings/docker.py:
```diff
@@ -0,0 +1,259 @@
+            'ENGINE': 'django.db.backends.postgresql_psycopg2',
+            'NAME': 'rtd',
+            'USER': 'docs',
+            'PASSWORD': 'docs',
```
Occhio alla password

yakky replied:
qui sarei quasi per togliere del tutto la password, perché ê quella interna ai container

yakky replied:
tolta
comunque è dentro il container, quindi non ha comunque molto senso avere una password

francescozaia reviewed (commented):
Come sapete non è prettamente il mio campo, per cui scusate se alcuni commenti sono fuori luogo.

yakky committed:
Remove comments

yakky committed:
Remove comments

LucaPrete commented on .dockerignore:
```diff
@@ -0,0 +1,13 @@
+.pytest
+deploy
+docs
+Dockerfile
```
Remove
* dockerfile
* .tox
* ....everything that is in .gitignore

LucaPrete replied:
Add restore_db.sh

yakky replied:
:thinking: I get file from gitignore in the docker image (eg: .tox)

yakky replied:
I pushed the current configuration: by `.tox`, `private*` etc they are copied in the container

LucaPrete commented on Dockerfile:
```diff
@@ -0,0 +1,33 @@
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN python -mvenv /virtualenv
```
If you can, try to remove

LucaPrete commented on Dockerfile:
```diff
@@ -0,0 +1,33 @@
+
+RUN python -mvenv /virtualenv
+
+RUN apt update && apt dist-upgrade -y
```
I would avoid to do apt dist-upgrade

yakky replied:
Removed

yakky replied:
Removed

LucaPrete commented on Dockerfile:
```diff
@@ -0,0 +1,33 @@
+ENV DJANGO_SETTINGS_MODULE=readthedocs.docsitalia.settings.docker
+ENV DEBUG=1
+WORKDIR /app
+RUN chmod +x docker/*sh
```
add default CMD ["/bin/bash"]

yakky replied:
Done

I also updated the dockerfile to a multi-staged build to test custom image for local build

LucaPrete commented on docker-compose.yml:
```diff
@@ -0,0 +1,95 @@
+version: '3.7'
+services:
+  http:
+    image: nginx:latest
```
please, always use tagged images

yakky replied:
done

LucaPrete commented on docker-compose.yml:
```diff
@@ -0,0 +1,95 @@
+  web:
+    hostname: web
+    build: .
+    image: &img web
```
I would call the image italia/docs.italia.it:latest (so you're more comfortable if you need to push it to dockerhub)

yakky replied:
done

LucaPrete commented on docker-compose.yml:
```diff
@@ -0,0 +1,95 @@
+    env_file: .env
+
+  api:
+    hostname: api
```
I would add container_name: name

yakky replied:
done

LucaPrete commented on docker/celery.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
```
#!/bin/env sh

yakky replied:
done

yakky committed:
Edit as per review

yakky committed:
Edit as per review

Commit:
Fix local docker compose setup

Commit:
Fix local docker compose setup

Commit:
Automate analysis-icu Elasticsearch plugin installation.

Commit:
Automate analysis-icu Elasticsearch plugin installation.

Commit:
Add .secrets env file in docker compose

Commit:
Add .secrets env file in docker compose

Commit:
Move .env to .dockerenv

Commit:
Move .env to .dockerenv

Commit:
Restore conf.py.tmpl and python_environments.py

Commit:
Restore conf.py.tmpl and python_environments.py

Commit:
Add Elasticsearch provisioning to web containers.
Provision Elasticsearch only if no index exists yet.

Commit:
Add Elasticsearch provisioning to web containers.
Provision Elasticsearch only if no index exists yet.

Commit:
Run tests inside docker container.

Commit:
Run tests inside docker container.

yakky committed:
Refactor docker file / docker compose to reuse layers

yakky committed:
Refactor docker file / docker compose to reuse layers

Commit:
Fix broken test

Commit:
Fix broken test

Commit:
Update Dockerfile

Commit:
Update Dockerfile

Commit:
Add ability to restore compressed database dump

Commit:
Add ability to restore compressed database dump

Commit:
Rename container name in docker compose test

Commit:
Rename container name in docker compose test

Commit:
Remove unused python3 packages in docker

Commit:
Remove unused python3 packages in docker

callmesangio committed:
Change postgres port to 5432 in test settings (#278)

* Change postgres port to 5432.

Commit:
Merge branch 'feature/docker' of github.com:italia/docs.italia.it into feature/docker

yakky committed:
Fix linting

yakky committed:
Remove use-wheel with comment

yakky committed:
Add comments to Dockerfile

yakky committed:
Fix linting

francescozaia committed:
Update Dockerfile

yakky committed:
Fix installed packages for celery-build image

yakky committed:
Limit ES memory consumption

yakky committed:
Set fixed path for better compatibility

yakky committed:
Try to make sure api image runs when ES is available

This needs to be fixed with proper polling cycle in startup script but we can address this later
